src/python/pants/engine/rules.py

```diff
@@ -26,14 +27,105 @@
 
 
 class _RuleVisitor(ast.NodeVisitor):
-  def __init__(self):
+  """Pull `Get` calls out of an @rule body and validate `yield` statements."""
+
+  def __init__(self, func, frame, parents_table):
     super(_RuleVisitor, self).__init__()
     self.gets = []
+    self._func = func
+    self._frame = frame
+    self._parents_table = parents_table
+    self._yields_in_assignments = set()
+
+  class YieldVisitError(Exception):
+    def __init__(self, node, func, frame, msg, *args, **kwargs):
+      filename, line_number, _, context_lines, _ = inspect.getframeinfo(frame, context=4)
+      err_msg = ("""\
+In function {func_name}: {msg}
+{filename}:{line_number}:
+{context_lines}
+
+The invalid `yield` statement (in the body of the above function) was: {node}
+""".format(func_name=func.__name__, msg=msg,
+           filename=filename, line_number=line_number,
```

Comment:
I think it'd make sense to use a calculated lineno taken from the node here instead of using the one for the annotation. From looking at ast's docs, I think we should be able to get a lineno from nodes.

Reply:
It's highly nontrivial to do this, as I tried to describe in a [previous comment](https://github.com/pantsbuild/pants/pull/7019#discussion_r245098989). I have a branch which does do this using a thirdparty library which I am planning to push after this passes CI: see https://github.com/cosmicexplorer/pants/compare/validate-yield-returns-in-rule-bodies...cosmicexplorer:import-asttokens-3rdparty-for-@rule-parse-errors?expand=1.